File: leetcode/py/p36-valid-sudoku.py
# LeetCode 36. Valid Sudoku
from typing import List

import pyperf


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        col_vals = [set() for _ in range(9)]
        sq_vals = [set() for _ in range(9)]
        for y in range(9):
            row = board[y]
            row_vals = set()
            for x in range(9):
                n = row[x]
                if n == ".":
                    continue
                if n in row_vals:
                    return False
                row_vals.add(n)
                if n in col_vals[x]:
                    return False
                col_vals[x].add(n)
                sq = (y // 3) * 3 + x // 3
                if n in sq_vals[sq]:
                    return False
                sq_vals[sq].add(n)
        return True


if __name__ == "__main__":
    tests = (
        (
            [
                ["5", "3", ".", ".", "7", ".", ".", ".", "."],
                ["6", ".", ".", "1", "9", "5", ".", ".", "."],
                [".", "9", "8", ".", ".", ".", ".", "6", "."],
                ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
                ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
                ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
                [".", "6", ".", ".", ".", ".", "2", "8", "."],
                [".", ".", ".", "4", "1", "9", ".", ".", "5"],
                [".", ".", ".", ".", "8", ".", ".", "7", "9"],
            ],
            True,
        ),
        (
            [
                ["8", "3", ".", ".", "7", ".", ".", ".", "."],
                ["6", ".", ".", "1", "9", "5", ".", ".", "."],
                [".", "9", "8", ".", ".", ".", ".", "6", "."],
                ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
                ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
                ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
                [".", "6", ".", ".", ".", ".", "2", "8", "."],
                [".", ".", ".", "4", "1", "9", ".", ".", "5"],
                [".", ".", ".", ".", "8", ".", ".", "7", "9"],
            ],
            False,
        ),
    )
    solution = Solution()
    for board, exp in tests:
        act = solution.isValidSudoku(board)
        if exp != act:
            print(f"Fail. board: {board}, exp: {exp}, act: {act}")

    # board = tests[0][0]
    # runner = pyperf.Runner()
    # runner.bench_func("f", solution.isValidSudoku, board)

File: leetcode/cpp/p36-valid-sudoku.cpp
// LeetCode 36. Valid Sudoku.
#include <iostream>
#include <unordered_set>
#include <vector>

#include "leetcode.hpp"

class Solution {
public:
    bool isValidSudoku(const std::vector<std::vector<char>>& board)
    {
        // Rows
        for (const auto& row : board) {
            std::unordered_set<char> row_vals;
            for (const auto n : row) {
                if (n != '.') {
                    if (row_vals.find(n) != row_vals.end()) {
                        return false;
                    }
                    row_vals.insert(n);
                }
            }
        }
        // Columns
        {
            std::unordered_set<char> col_vals[9] = {};
            for (const auto& row : board) {
                for (size_t x = 0; x < 9; ++x) {
                    const auto n = row[x];
                    if (n != '.') {
                        if (col_vals[x].find(n) != col_vals[x].end()) {
                            return false;
                        }
                        col_vals[x].insert(n);
                    }
                }
            }
        }
        // Squares
        {
            std::unordered_set<char> sq_vals[9] = {};
            for (size_t y = 0; y < 9; ++y) {
                const auto& row = board[y];
                for (size_t x = 0; x < 9; ++x) {
                    const auto n = row[x];
                    const size_t sq = (y / 3) * 3 + x / 3;
                    if (n != '.') {
                        if (sq_vals[sq].find(n) != sq_vals[sq].end()) {
                            return false;
                        }
                        sq_vals[sq].insert(n);
                    }
                }
            }
        }
        return true;
    }
};

int main()
{
    struct TestCase {
        std::vector<std::vector<char>> board;
        bool exp;
    };
    const TestCase tcs[] = {
        { { { '5', '3', '.', '.', '7', '.', '.', '.', '.' }, { '6', '.', '.', '1', '9', '5', '.', '.', '.' }, { '.', '9', '8', '.', '.', '.', '.', '6', '.' }, { '8', '.', '.', '.', '6', '.', '.', '.', '3' }, { '4', '.', '.', '8', '.', '3', '.', '.', '1' }, { '7', '.', '.', '.', '2', '.', '.', '.', '6' }, { '.', '6', '.', '.', '.', '.', '2', '8', '.' }, { '.', '.', '.', '4', '1', '9', '.', '.', '5' }, { '.', '.', '.', '.', '8', '.', '.', '7', '9' } },
            true },
        { { { '8', '3', '.', '.', '7', '.', '.', '.', '.' }, { '6', '.', '.', '1', '9', '5', '.', '.', '.' }, { '.', '9', '8', '.', '.', '.', '.', '6', '.' }, { '8', '.', '.', '.', '6', '.', '.', '.', '3' }, { '4', '.', '.', '8', '.', '3', '.', '.', '1' }, { '7', '.', '.', '.', '2', '.', '.', '.', '6' }, { '.', '6', '.', '.', '.', '.', '2', '8', '.' }, { '.', '.', '.', '4', '1', '9', '.', '.', '5' }, { '.', '.', '.', '.', '8', '.', '.', '7', '9' } },
            false }
    };
    Solution solution;
    for (const auto& tc : tcs) {
        const auto act = solution.isValidSudoku(tc.board);
        if (tc.exp != act) {
            std::cerr << "Fail. board: " << leetcode::to_string(tc.board)
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
    return 0;
}
